### PR TITLE
Use organisation API

### PR DESCRIPTION
**Pull request by jekuaitk**
Handles the upgrade to [OS2Forms Organisation API
](https://github.com/itk-dev/os2forms_organisation_api)

https://jira.itkdev.dk/browse/FORMS-985

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -9,6 +9,10 @@ Versioning](https://semver.org/spec/v2.0.0.html).
 
 ## [Unreleased]
 
+## [2.0.0] 2013-12-11
+
+- Upgraded to use [OS2Forms Organisation API](https://github.com/itk-dev/os2forms_organisation_api).
```
Comment (rimi-itk) [inline CHANGELOG.md]: The API has nothing to do with OS2Forms. It's a general API for organisation data.

File: CHANGELOG.md [hunk 2]
```diff
@@ -9,6 +9,10 @@ Versioning](https://semver.org/spec/v2.0.0.html).
 
 ## [Unreleased]
 
+## [2.0.0] 2013-12-11
```
Comment (rimi-itk) [inline CHANGELOG.md]: Is this (also) a release?

File: src/Form/SettingsForm.php [hunk 3]
```diff
@@ -16,13 +15,8 @@
 final class SettingsForm extends FormBase {
   use StringTranslationTrait;
 
-  public const AUTHORITY_CVR = 'authority_cvr';
-  public const CACHE_EXPIRATION = 'cache_expiration';
-  public const CERTIFICATE = 'certificate';
+  public const ORGANISATION_BASE_API_ENDPOINT = 'organisation_base_api_endpoint';
   public const TEST_MODE = 'test_mode';
```
Comment (rimi-itk) [inline src/Form/SettingsForm.php]: Does `TEST_MODE` make sense anymore? I suppose that the API endpoint must point to a test API if we need to run in test mode.

File: src/Form/SettingsForm.php [hunk 4]
```diff
@@ -71,120 +56,13 @@ public function getFormId() {
...
-    ];
-
-    $organisationServiceEndpoint = $this->settings->getOrganisationServiceEndpoint();
-    $form[self::ORGANISATION_SERVICE_ENDPOINT_REFERENCE] = [
+    $organisationApiEndpoint = $this->settings->getOrganisationApiEndpoint();
+    $form[self::ORGANISATION_BASE_API_ENDPOINT] = [
       '#type' => 'textfield',
-      '#title' => $this->t('Organisation service endpoint reference'),
+      '#title' => $this->t('Organisation base API endpoint'),
+      '#description' => $this->t('Endpoint to local docker container, i.e. http://host.docker.internal:PORT/api/v1/'),
```
Comment (rimi-itk) [inline src/Form/SettingsForm.php]: You should not mention docker here. It's just an API endpoint and no assumptions on how that API is implemented should be made here.

File: src/Helper/OrganisationApiHelper.php [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+   * Constructor.
+   */
+  public function __construct(Settings $settings) {
+    $this->settings = $settings;
+  }
+
+  /**
+   * Get curl client.
+   */
+  private function getCurlClient(): CurlHttpClient {
```
Comment (rimi-itk) [inline src/Helper/OrganisationApiHelper.php]: Consider using https://www.drupal.org/docs/contributed-modules/http-client-manager/introduction.

File: src/Helper/OrganisationApiHelper.php [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+   * @phpstan-return array<string, mixed>
+   */
+  public function getBrugerInformationer(string $brugerId): array {
+    $client = $this->getCurlClient();
+
+    $this->settings->getOrganisationApiEndpoint();
+
+    $response = $client->request('GET', $this->settings->getOrganisationApiEndpoint() . 'bruger/' . $brugerId);
+
+    if (200 === $response->getStatusCode()) {
```
Comment (rimi-itk) [inline src/Helper/OrganisationApiHelper.php]:
```suggestion
    if (Response::HTTP_OK === $response->getStatusCode()) {
```

Comment (rimi-itk) [inline src/Helper/OrganisationApiHelper.php]: Is an exception thrown if the status code is not 200?

File: src/Helper/OrganisationApiHelper.php [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+   * Get bruger informationer for bruger.
+   *
+   * @phpstan-return array<string, mixed>
+   */
+  public function getBrugerInformationer(string $brugerId): array {
+    $client = $this->getCurlClient();
+
+    $this->settings->getOrganisationApiEndpoint();
+
+    $response = $client->request('GET', $this->settings->getOrganisationApiEndpoint() . 'bruger/' . $brugerId);
```
Comment (rimi-itk) [inline src/Helper/OrganisationApiHelper.php]: This code is (almost) duplicated multiple times. Create a helper function, `get`, say, to keep things [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself):

```php
private function get(string $path): Response
{
  return $this->getClient()->request('GET', $this->settings->getOrganisationApiEndpoint() . $path);
}
```

File: src/Helper/OrganisationApiHelper.php [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+
+    if (200 === $response->getStatusCode()) {
+      return $response->toArray();
+    }
+    else {
+      return [];
+    }
```
Comment (rimi-itk) [inline src/Helper/OrganisationApiHelper.php]: You don't need the `else` branch:

```suggestion
    }

    return [];
```

File: src/Helper/OrganisationApiHelper.php [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+        return reset($managerIds);
+      }
+      else {
+        return '';
+      }
+
+    }
+    else {
+      return '';
+    }
```
Comment (rimi-itk) [inline src/Helper/OrganisationApiHelper.php]: Get rid of the `else` branches.

```suggestion
    }
    
    return '';
```

Consider telling why is makes sense to return an empty string when the request fails.

File: src/Helper/Settings.php [hunk 10]
```diff
@@ -34,55 +34,10 @@ public function __construct(KeyValueFactoryInterface $keyValueFactory) {
...
-    return $this->get(SettingsForm::ORGANISATION_TEST_LEDER_ROLLE_UUID, '');
-  }
-
-  /**
-   * Get organisation production manager role id.
-   */
-  public function getOrganisationProductionManagerRoleId(): string {
-    return $this->get(SettingsForm::ORGANISATION_PROD_LEDER_ROLLE_UUID, '');
+  public function getOrganisationApiEndpoint(): string {
+    return $this->get(SettingsForm::ORGANISATION_BASE_API_ENDPOINT, '');
```
Comment (rimi-itk) [inline src/Helper/Settings.php]: Why is the constant called `ORGANISATION_BASE_API_ENDPOINT` when it's just an organisation API endpoint (without “base”)?

**Review by rimi-itk — CHANGES_REQUESTED** [review]

File: CHANGELOG.md (hunk 2 above)
Comment (jekuaitk) [reply CHANGELOG.md]: I suppose it should be tested first. Removed for now.

File: CHANGELOG.md (hunk 1 above)
Comment (jekuaitk) [reply CHANGELOG.md]: Rephrased.

File: src/Form/SettingsForm.php (hunk 3 above)
Comment (jekuaitk) [reply src/Form/SettingsForm.php]: It does not make sense anymore. Removed.

File: src/Form/SettingsForm.php (hunk 4 above)
Comment (jekuaitk) [reply src/Form/SettingsForm.php]: Rephrased.

File: src/Helper/OrganisationApiHelper.php (hunk 6 above)
Comment (jekuaitk) [reply src/Helper/OrganisationApiHelper.php]: Updated.

File: src/Helper/Settings.php (hunk 10 above)
Comment (jekuaitk) [reply src/Helper/Settings.php]: Updated.

File: src/Helper/OrganisationApiHelper.php (hunk 5 above)
Comment (jekuaitk) [reply src/Helper/OrganisationApiHelper.php]: Used HTTP Client Manager.

File: src/Helper/OrganisationApiHelper.php (hunk 7 above)
Comment (jekuaitk) [reply src/Helper/OrganisationApiHelper.php]: Refactored code.

File: src/Helper/OrganisationApiHelper.php (hunk 8 above)
Comment (jekuaitk) [reply src/Helper/OrganisationApiHelper.php]: Refactored code.

File: src/Helper/OrganisationApiHelper.php (hunk 9 above)
Comment (jekuaitk) [reply src/Helper/OrganisationApiHelper.php]: Refactored code.

File: src/Helper/OrganisationApiHelper.php (hunk 6 above)
Comment (jekuaitk) [reply src/Helper/OrganisationApiHelper.php]: Refactored code.

File: src/Exception/ApiException.php [hunk 11]
```diff
@@ -0,0 +1,10 @@
+<?php
+
+namespace Drupal\os2forms_organisation\Exception;
+
+/**
+ * Invalid argument exception.
```
Comment (rimi-itk) [inline src/Exception/ApiException.php]: It looks more like an API exception.

File: src/Form/SettingsForm.php [hunk 12]
```diff
@@ -25,11 +25,19 @@ final class SettingsForm extends FormBase {
...
+   *
+   * @var \Drupal\os2forms_organisation\Helper\OrganisationApiHelper
+   */
+  private OrganisationApiHelper $helper;
+
   /**
    * Constructor.
    */
-  public function __construct(Settings $settings) {
+  public function __construct(Settings $settings, OrganisationApiHelper $helper) {
```
Comment (rimi-itk) [inline src/Form/SettingsForm.php]: Use https://www.php.net/manual/en/language.oop5.decon.php#language.oop5.decon.constructor.promotion here.

File: src/Helper/OrganisationApiHelper.php [hunk 13]
```diff
@@ -26,167 +30,166 @@ class OrganisationApiHelper {
...
-  private function getCurlClient(): CurlHttpClient {
-    if ($this->client) {
-      return $this->client;
-    }
-    else {
-      $this->client = new CurlHttpClient();
-
-      return $this->client;
-    }
+    $this->httpClient = $httpClient;
```
Comment (rimi-itk) [inline src/Helper/OrganisationApiHelper.php]: Use https://www.php.net/manual/en/language.oop5.decon.php#language.oop5.decon.constructor.promotion here.

File: src/Helper/OrganisationApiHelper.php [hunk 14]
```diff
@@ -26,167 +30,166 @@ class OrganisationApiHelper {
...
    * Search for bruger.
    *
    * @phpstan-return array<string, mixed>
+   *
+   * @throws \Drupal\os2forms_organisation\Exception\ApiException
+   *   API exception.
    */
   public function searchBruger(string $query): array {
-    $client = $this->getCurlClient();
+    $response = $this->get('bruger?page=1&navn=' . $query);
```
Comment (rimi-itk) [inline src/Helper/OrganisationApiHelper.php]: You should not manually build the query string. Pass an array with the parameters to the `get` function:

```suggestion
    $response = $this->get('bruger', ['page' => 1, 'navn' => $query]);
```

File: src/Helper/OrganisationApiHelper.php [hunk 15]
```diff
@@ -26,167 +30,166 @@ class OrganisationApiHelper {
...
-    $this->settings->getOrganisationApiEndpoint();
+    $response = $this->get('bruger/' . $brugerId);
 
-    $response = $client->request('GET', $this->settings->getOrganisationApiEndpoint() . 'bruger/' . $brugerId);
-
-    if (200 === $response->getStatusCode()) {
-      return $response->toArray();
-    }
-    else {
+    if (Response::HTTP_OK != $response->getStatusCode()) {
```
Comment (rimi-itk) [inline src/Helper/OrganisationApiHelper.php]: Will this ever happen (without an error being thrown in `get`)?

File: src/Plugin/WebformElement/MineOrganisationsData.php [hunk 16]
```diff
@@ -643,9 +649,18 @@ private function getSearchResult(array &$element, string $query): array {
...
 
-    $brugere = $this->organisationHelper->searchBruger($query);
+    try {
+      $brugere = $this->organisationHelper->searchBruger($query);
+    }
+    catch (ApiException $e) {
+      $brugere = NULL;
+    }
+
+    if (NULL === $brugere) {
```
Comment (rimi-itk) [inline src/Plugin/WebformElement/MineOrganisationsData.php]:
```suggestion
```

File: src/Plugin/WebformElement/MineOrganisationsData.php [hunk 17]
```diff
@@ -643,9 +649,18 @@ private function getSearchResult(array &$element, string $query): array {
    */
   private function getSearchUserIds(string $query): array {
 
-    $brugere = $this->organisationHelper->searchBruger($query);
+    try {
+      $brugere = $this->organisationHelper->searchBruger($query);
+    }
+    catch (ApiException $e) {
+      $brugere = NULL;
```
Comment (rimi-itk) [inline src/Plugin/WebformElement/MineOrganisationsData.php:657]: Consider logging the exception.

File: src/Plugin/WebformElement/MineOrganisationsData.php [hunk 18]
```diff
@@ -643,9 +649,18 @@ private function getSearchResult(array &$element, string $query): array {
...
+    }
+
+    if (NULL === $brugere) {
+      $this->messenger()->addMessage($this->t('Could not fetch organisation data. Contact form owner.'));
+    }
 
-    return array_map(static fn($value) => $value['id'], $brugere);
+    return array_map(static fn($value) => $value['id'], $brugere ?? []);
```
Comment (rimi-itk) [inline src/Plugin/WebformElement/MineOrganisationsData.php]: https://www.php.net/manual/en/function.array-column.php is your friend.

**Review by rimi-itk — CHANGES_REQUESTED** [review]

File: src/Helper/OrganisationApiHelper.php (hunk 14 above)
Comment (jekuaitk) [reply src/Helper/OrganisationApiHelper.php]: Used `get` to build query string.

File: src/Helper/OrganisationApiHelper.php (hunk 13 above)
Comment (jekuaitk) [reply src/Helper/OrganisationApiHelper.php]: Used constructor promotion.

File: src/Form/SettingsForm.php (hunk 12 above)
Comment (jekuaitk) [reply src/Form/SettingsForm.php]: Used constructor promotion.

File: src/Exception/ApiException.php (hunk 11 above)
Comment (jekuaitk) [reply src/Exception/ApiException.php]: Updated.

File: src/Plugin/WebformElement/MineOrganisationsData.php (hunk 16 above)
Comment (jekuaitk) [reply src/Plugin/WebformElement/MineOrganisationsData.php]: 👍

File: src/Plugin/WebformElement/MineOrganisationsData.php (hunk 17 above)
Comment (jekuaitk) [reply src/Plugin/WebformElement/MineOrganisationsData.php:657]: Considered and logged.

File: src/Plugin/WebformElement/MineOrganisationsData.php (hunk 18 above)
Comment (jekuaitk) [reply src/Plugin/WebformElement/MineOrganisationsData.php]: 👍

File: src/Helper/OrganisationApiHelper.php (hunk 15 above)
Comment (jekuaitk) [reply src/Helper/OrganisationApiHelper.php]: It doesn't appear so. Removed.

**Review by rimi-itk — APPROVED** [review]

**Pull request merged**